### PR TITLE
New lmi path

### PR DESCRIPTION
**Pull request by thesiv**
The questions:
1. Should I split (because it also contains `fs::{i,o}fstream`) or rename (to `filesystem.hpp`) `path.hpp` ?
1. Should functions which returns or takes as a parameter `std::string` typed path be rewritten?
   ```
   std::string AddDataDir(std::string const& a_filename)
   void LMI_SO PrintRosterHeaders     (               std::string const& file_name);
   void LMI_SO PrintRosterTabDelimited(Ledger const&, std::string const& file_name);
   void global_settings::set_data_directory(std::string const& s)
   ```
1. Should I add the unit test for `fs::path` and `fs::{i,o}fstream`? Currently the separate tests are not added because the main difference is UTF8 encoding and to test it properly `c.uft-8` locale should be set, but it's not fully supported under windows. And other things like `fs::path::operator<<` are tested in `authenticity_test.cpp` and `path_utility_test.cpp`.

### CODE REVIEW

**Comment by thesiv** [comment]
Replaces https://github.com/vadz/lmi/pull/149.

File: authenticity_test.cpp [hunk 1]
```diff
@@ -416,7 +393,7 @@ void PasskeyTest::TestPasskey() const
     BOOST_TEST(!fs::exists("passkey"));
     Authenticity::ResetCache();
     BOOST_TEST_EQUAL
-        ("Unable to read passkey file 'passkey'. Try reinstalling."
+        ("Unable to read passkey file './passkey'. Try reinstalling."
```
Comment (vadz) [inline authenticity_test.cpp:396]: This looks like a problem. I.e. the previous message seems better, can we avoid the leading `./` here (and below, of course)?

File: path.hpp [hunk 2]
```diff
@@ -0,0 +1,221 @@
+// The path class.
+//
+// Copyright (C) 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020 Gregory W. Chicares.
```
Comment (vadz) [inline path.hpp:3]:
```suggestion
// Copyright (C) 2020 Gregory W. Chicares.
```
(AFAICS this is a new file)

File: path.cpp [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+// snail: Chicares, 186 Belle Woods Drive, Glastonbury CT 06033, USA
+
+#include "pchfile.hpp"
+
+#include "path.hpp"
+
+namespace fs
+{
+
+path::path(const path& p)
```
Comment (vadz) [inline path.cpp:29]: I'd implement this and plenty of other trivial functions inline.

In fact, I wonder if we need `path.cpp` at all, it looks like we could reasonably make everything inline and then avoid all the changes to the makefiles. Why did you decide to add it, is there some advantage to having it I'm missing?

File: path.hpp [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+    path& operator=(std::string const& source);
+    path& operator=(char const* source);
+ 
+    // Appends.
+    path& operator/=(path const& p);
+    path& operator/=(std::string const& source);
+ 
+    // Modifiers.
+    void  clear() noexcept;
+    path& make_preferred(); // This used only in native_path().
```
Comment (vadz) [inline path.hpp:88]:
```suggestion
    path& make_preferred(); // This is used only in native_path().
```

File: path.hpp [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+    path& remove_filename();
+    path& replace_filename(const path& replacement);
+    path& replace_extension(const path& replacement = path());
+    void  swap(path& rhs) noexcept;
+ 
+    // Native format observer.
+    operator std::filesystem::path() const;
+
+    // The function used only in native_path(). Returns the UFT8 string
```
Comment (vadz) [inline path.hpp:97]:
```suggestion
    // The function is used only in native_path(). Returns the UFT8 string
```

File: path.hpp [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+    // with the native separators ('\\' under Windows).
+    // Note that the corresponding function doesn't exists in
```
Comment (vadz) [inline path.hpp:99]:
```suggestion
    // Note that the corresponding function doesn't exist in
```

File: path.hpp [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+// Implement own functions, which return the path.
+
+path LMI_SO absolute(const path& p);
+path LMI_SO absolute(const path& p, std::error_code& ec);
+path LMI_SO current_path();
+path LMI_SO current_path(std::error_code& ec);
+
+/// File streams.
+///
+/// Motivation: implement own file streams, which constructs from the own
```
Comment (vadz) [inline path.hpp:167]:
```suggestion
/// Motivation: our own file streams can be constructed from our own
```

File: path.hpp [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+/// path class. fs::path has `operator std::filesystem::path()`, but
+/// the following code fails:
+///
+/// fs::path p{"file.ext"};
+/// fs::ifstream s{p};
+///
+/// So implement the own classes for the exact match of constructor parameters.
```
Comment (vadz) [inline path.hpp:174]:
```suggestion
/// So implement our own classes for the exact match of constructor parameters.
```

File: path.hpp [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+///
+/// Note that fs::path constructs from a string using fs::u8path() which leads
```
Comment (vadz) [inline path.hpp:176]:
```suggestion
/// Note that fs::path can be constructed from a string using fs::u8path() which leads
```

**Review by vadz — COMMENTED** [review]
I didn't have time to look at 90% of the changes yet, but I already have a question after just a quick look: could we have `fs::path::string()` that would return `std::filesystem::generic_u8string()`? It seems that this would allow us to avoid making plenty of changes replacing `string()` calls with `generic_u8string()` which should drastically reduce the size of the diff. And, of course, `string()` is much nicer/simpler to use too. And IMO we should make the short and simple function return what we almost always want.

Any reason not to do it like this?

**Comment by thesiv** [comment]
Fixed.

**Pull request closed without merging**